File: app/src/main/java/com/afterlogic/aurora/drive/data/common/util/MultiApiUtil.java
package com.afterlogic.aurora.drive.data.common.util;

import com.afterlogic.aurora.drive.core.consts.Const;
import com.afterlogic.aurora.drive.data.common.network.SessionManager;
import com.afterlogic.aurora.drive.model.AuroraSession;

import javax.inject.Provider;

/**
 * Created by sashka on 03.02.17.<p/>
 * mail: sunnyday.development@gmail.com
 */

public class MultiApiUtil {

    public static <T> T chooseByApiVersion(SessionManager sessionManager, Provider<T> p7, Provider<T> p8){
        AuroraSession session = sessionManager.getSession();

        if (session == null) {
            throw new IllegalStateException("Not authorized. So can't provide any.");
        }

        switch (session.getApiVersion()){
            case Const.ApiVersion.API_P7:
                return p7.get();
            case Const.ApiVersion.API_P8:
                return p8.get();
            default:
                return null;
        }
    }
}

File: app/src/main/java/com/afterlogic/aurora/drive/data/common/multiapi/MultiApiUtil.kt
package com.afterlogic.aurora.drive.data.common.multiapi

import com.afterlogic.aurora.drive.core.consts.Const
import com.afterlogic.aurora.drive.data.common.network.SessionManager

import javax.inject.Provider

/**
 * Created by sashka on 03.02.17.
 *
 *
 * mail: sunnyday.development@gmail.com
 */

object MultiApiUtil {

    @Throws(MultiApiServiceError::class)
    fun <T> chooseByApiVersion(sessionManager: SessionManager,
                               p7: Provider<T>, p8: Provider<T>): T {

        val session = sessionManager.session
                ?: throw MultiApiServiceNotAuthorizedError()

        return when (session.apiVersion) {
            Const.ApiVersion.API_P7 -> p7.get()
            Const.ApiVersion.API_P8 -> p8.get()
            else -> throw MultiApiServiceUnknownApiError()
        }

    }

}
